garage/np/algos/batch_polopt.py

```diff
@@ -1,135 +1,84 @@
-                logger.log('Saving snapshot...')
-                params = self.get_itr_snapshot(itr, samples_data)
-                self.current_itr = itr + 1
-                params['algo'] = self
-                if self.store_paths:
-                    params['paths'] = samples_data['paths']
-                snapshotter.save_itr_params(itr, params)
-                logger.log('saved')
-                logger.log(tabular)
-                if self.plot:
-                    plotter.update_plot(self.policy, self.max_path_length)
-                    if self.pause_for_plot:
-                        input('Plotting evaluation run: Press Enter to '
-                              'continue...')
-
-        plotter.close()
-        self.shutdown_worker()
-
-    def log_diagnostics(self, paths):
-        self.policy.log_diagnostics(paths)
-        self.baseline.log_diagnostics(paths)
-
-    def init_opt(self):
-        """
-        Initialize the optimization procedure. If using tf, this may
-        include declaring all the variables and compiling functions
-        """
-        raise NotImplementedError
+        self.discount = discount
+        self.max_path_length = max_path_length
 
-    def get_itr_snapshot(self, itr, samples_data):
-        """
-        Returns all the data that should be saved in the snapshot for this
-        iteration.
+        self.episode_reward_mean = deque(maxlen=100)
+
+    def process_samples(self, itr, paths):
+        """Return processed sample data based on the collected paths.
+        This method is currently used by cem and cam_es.
+
+        Args:
+            itr (int): Iteration number.
+            paths (list[dict]): A list of collected paths
+
+        Returns:
+            dict: Processed sample data, with key
+                * average_return: (float)
         """
-        raise NotImplementedError
+        baselines = []
+        returns = []
+
+        if hasattr(self.baseline, 'predict_n'):
+            all_path_baselines = self.baseline.predict_n(paths)
+        else:
+            all_path_baselines = [
+                self.baseline.predict(path) for path in paths
+            ]
+
+        for idx, path in enumerate(paths):
+            # baselines
+            path['baselines'] = all_path_baselines[idx]
+            baselines.append(path['baselines'])
+
+            # returns
+            path['returns'] = special.discount_cumsum(path['rewards'],
+                                                      self.discount)
+            returns.append(path['returns'])
+
+        undiscounted_returns = [sum(path['rewards']) for path in paths]
+
+        samples_data = dict(average_return=np.mean(undiscounted_returns))
+
+        tabular.record('Iteration', itr)
+        tabular.record('AverageReturn', np.mean(undiscounted_returns))
+
+        return samples_data
 
-    def optimize_policy(self, itr, samples_data):
-        raise NotImplementedError
+    def train_once(self, itr, paths):
```

Comment:
make BatchPolopt an ABC as well, so you can declare this method abstract (or skip implementing it--idk what the rules are for one ABC inheriting from another)